File: leetcode/29.java
class Solution {
    /**
     * 两数相除
     * 
     * dividend 被除数
     * divisor 除数
     */
     // 32位  -2^31 = 2^31 - 1
     

     // 举个例子 4位bit  0000  第一位表示 +-

     // 2^1  =  10   2^0 = 01
     
     // 111 - 000 ---->   -8 -> 7   -> 4位  --> 2^3-1 ~ -2^3

     // 
     public int divide(int dividend, int divisor) {
        // 处理整型溢出问题，-2³¹=0x80000000
        // -2^31 / -1 =2^31   2^31 溢出了 > 2^31 - 1
        if (dividend == -0x80000000 && divisor == -1) {
            return 0x80000000 - 1;
        }

        // 用来记录除数和被除数是负整数的个数，初始化为2
        int negative = 2;

        // 如果被除数是正整数，则转换成对应的负整数
        if (dividend > 0) {
            negative--;
            dividend = -dividend;
        }

        // 如果除数是正整数，则转换成对应的负整数
        if (divisor > 0) {
            negative--;
            divisor = -divisor;
        }
        // 两个都转成负数

        int result = divideCore(dividend, divisor);
        return negative == 1 ? -result : result;
    }

    /**
     * 使用减法实现两个负整数的相除操作
     * 
     * dividend：被除数
     * divisor：除数
     */
    
    // -100  -9
        
    private int divideCore(int dividend, int divisor) {
        int ans = 0;
        while (dividend <= divisor) {
            // 商
            int quotient = 1;
            
            // 临时变量
            int value = divisor;

            // 注意题目要求，只能使用加法和减法
            // 因此，通过value+value实现value*2的效果
            // 注意这里 value 是大于等于 0xc0000000，不是小于等于，因为是负数，思维得转变过来
            while (value >= 0xc0000000 && dividend <= value + value) {
                // quotient 通过自身不断叠加实现乘以 2 的效果
                quotient += quotient;  // 1 -> 2 -> 4 -> 8
                // value 通过自身不断叠加实现乘以 2 的效果
                value += value; // -9 -> -18 -> -36 -> -72 ->  // -144
            }

            ans += quotient;
            dividend -= value;
        }

        return ans;
    }
}

File: leetcode/29.js
// 两数相除
// 解题思路
// https://leetcode.cn/problems/divide-two-integers/solution/by-i3lissful-devvdney-mkao/

// 100  -9
// 100  -11   1

// 32 位  -2^31 ~ 2^31 -1
function divide(dividend, divisor) {
  // 整型溢出
  if (dividend == -0x80000000 && divisor == -1) {
    return 0x80000000 - 1
  }
  if (divisor == 0) {
    return 0x80000000 - 1
  }

  // 不能转成用整数求， -2^31 会溢出
  let nums = 2
  if (dividend > 0) {
    nums--
    dividend = -dividend
  }
  if (divisor > 0) {
    nums--
    divisor = -divisor
  }

  let val = divideCode2(dividend, divisor)

  return nums === 1 ? -val : val
}

// -100、 -9
function divideCode(dividend, divisor) {
  let ans = 0
  // -11      -9
  while (dividend <= divisor) {
    ans++
    dividend = dividend - divisor
  }

  return ans
}

//
console.log(divide(18, -9)) // 11

// 优化算法 加入指数

function divideCode2(dividend, divisor) {
  let ans = 0
  // -100      -9
  while (dividend <= divisor) {
    let q = 1
    let value = divisor
    // 指数增长去算 * 2 倍 若 2倍的value > 除数 则停止
    while (dividend < value + value) {
      // * 2
      q = q + q
      value = value + value
    }
    dividend = dividend - value
    ans += q
  }

  return ans
}

// 优化算法  处理边界
// 如果数值是 divisor = 2^30 那在 2^30 * 2 = 2^31 会溢出

function divideCode3(dividend, divisor) {
  let ans = 0
  // -100      -9
  while (dividend <= divisor) {
    let q = 1
    let value = divisor
    // 指数增长去算 * 2 倍 若 2倍的value > 除数 则停止
    while (value > -Math.pow(2, 30) && dividend < value + value) {
      // * 2
      q = q + q
      value = value + value
    }
    dividend = dividend - value
    ans += q
  }

  return ans
}
